proposals/base-relative-assets.md

```diff
@@ -0,0 +1,132 @@
+
+People have tried to use `./` in the `base` option to make this happen to, such as in [this issue](https://github.com/withastro/astro/issues/4229).
+
+The primary use-case is wanting to deploy a site that might live under multiple subdomains. Using relative links makes it portable.
+
+# Detailed design
+
+This new configuration option will affect these types of assets:
+
+- Link tags added by us from bundled CSS.
+- Hoisted script tags added by us from bundled hoisted scripts.
+- Links to scripts added to `<astro-island>`s.
+
+Any other type of asset, such as images, are not controlled by Astro and will continue to not be so. The developer will need to make these links relative themselves. They can also use the `<base>` tag to make this easier.
+
+## Implementation
+
+During the build when we go to generate pages there are a few places where we resolve the final URL for links.
+
+- For link tags it is done [here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/build/generate.ts#L300).
+- For hoisted scripts it is [here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/build/generate.ts#L301)
+- For links added to astro-islands it [is here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/build/generate.ts#L352).
+
+We should generalize this logic so that there's one function that gets called to resolve the URL given the `linkStyle` option, as well as the base and subpath. That function will use the `linkStyle` config to determine which algorithm to use.
+
+For SSR the same things are done [here](https://github.com/withastro/astro/blob/c0cb65b18db888ed46fe9a0b7d66d062de245f5e/packages/astro/src/core/app/index.ts#L63) for example. Again, this should be consolidated so we're not doing the same sort of logic in many different places, creating areas for bugs. Instead SSG and SSR should use shared logic.
+
+## 404.astro
+
+If using `linkStyle: 'relative'` a 404.astro page __will not be allowed__. This is because it is impossible to make these links be relative in SSG mode. That's because a 404.astro might be used by paths such as:
+
+- /foo/
+- /foo/bar/
+- /foo/bar/baz/
+
+If we statically generate relative links then these will be incorrect. So we will throw an error if using a 404.astro page with `'relative'` and suggest using `'absolute'` instead.
+
+## Imported images
+
+Images that are imported will *not* be relative. That means this:
+
+```astro
+---
+import penguinUrl from '../assets/penguin.png';
+---
+
+<img src={penguinUrl} />
+```
+
+It might be *possible* to make this work through a Vite plugin but would be a bit of work and might be error prone. 
```

Comment:
If we set `base: './'` in Vite's config, this should work today so I'm not sure if this part is a problem unless I'm missing something.

As long as an asset is imported, Vite can generate a relative link dynamically based on the current chunk.

Reply:
@bluwy Relative to the file that's importing it though, right? Not relative to the page. Vite is not aware of the page that an import is being used on (it can be used on multiple pages!). So either way you as a user need to convert it to a relative link.

Even without this, we could have a directive like `relative:src` that uses the current URL and converts the provided absolute URL into a relative one.

Reply:
I did some test locally and seems like you might be right. Looks like Vite doesn't handle `base` in SSR at all too. Maybe Vite's SSR manifest or [advanced base option](https://vitejs.dev/guide/build.html#advanced-base-options) could help here, but this seems like a tricky one indeed.